File: go/pkg/hashing/hashing.go
// Package hashing implements the BLAKE3 multihash used by ArxDB.
//
// A Hash is a 34-byte multihash: 0x1e (BLAKE3 code) || 0x20 (32-byte length)
// || 32-byte BLAKE3 digest. The multihash prefix makes the hash algorithm
// self-describing, so a future algorithm migration does not break stored
// objects.
//
// This must produce byte-identical output to the Python reference
// (src/arxdb/storage/hashing.py) for the same input.
package hashing

import (
	"encoding/hex"
	"fmt"

	"github.com/zeebo/blake3"
)

// Multihash constants.
const (
	BLAKE3Code = 0x1E
	BLAKE3Len  = 0x20
	HashSize   = 34 // 1 (code) + 1 (len) + 32 (digest)
)

// Hash is a 34-byte BLAKE3 multihash.
type Hash [HashSize]byte

// HashBytes returns the 34-byte BLAKE3 multihash of data.
func HashBytes(data []byte) Hash {
	digest := blake3.Sum256(data)
	var h Hash
	h[0] = BLAKE3Code
	h[1] = BLAKE3Len
	copy(h[2:], digest[:])
	return h
}

// HashHex returns the hex string of the multihash of data.
func HashHex(data []byte) string {
	h := HashBytes(data)
	return hex.EncodeToString(h[:])
}

// FromHex reconstructs a Hash from its hex string.
func FromHex(hexstr string) (Hash, error) {
	var h Hash
	raw, err := hex.DecodeString(hexstr)
	if err != nil {
		return h, err
	}
	if len(raw) != HashSize {
		return h, fmt.Errorf("Hash must be %d bytes, got %d", HashSize, len(raw))
	}
	copy(h[:], raw)
	return h, nil
}

// IsValidHash reports whether b is a well-formed 34-byte BLAKE3 multihash.
func IsValidHash(b []byte) bool {
	return len(b) == HashSize && b[0] == BLAKE3Code && b[1] == BLAKE3Len
}

// Hex returns the hex string of the hash.
func (h Hash) Hex() string {
	return hex.EncodeToString(h[:])
}

// Bytes returns the hash as a byte slice.
func (h Hash) Bytes() []byte {
	return h[:]
}

File: src/arxdb/storage/hashing.py
"""Hashing — BLAKE3 multihash and the `Hash` type.

A `Hash` is a 34-byte multihash: `0x1e` (BLAKE3 code) ‖ `0x20` (32-byte length)
‖ 32-byte BLAKE3 digest. The multihash prefix makes the hash algorithm
self-describing, so a future algorithm migration does not break stored objects.

Public API (Phase 1):
    Hash            — bytes subclass carrying the 34-byte multihash
    hash_bytes(data: bytes) -> Hash
    hash_hex(data: bytes) -> str
    from_hex(hexstr: str) -> Hash
    is_valid_hash(h: bytes) -> bool
"""

from __future__ import annotations

import blake3

# Multihash constants
BLAKE3_CODE = 0x1E
BLAKE3_LEN = 0x20
HASH_SIZE = 34  # 1 (code) + 1 (len) + 32 (digest)


class Hash(bytes):
    """A 34-byte multihash (BLAKE3). Subclasses bytes for zero-copy interop."""

    def __new__(cls, raw: bytes) -> "Hash":
        if len(raw) != HASH_SIZE:
            raise ValueError(f"Hash must be {HASH_SIZE} bytes, got {len(raw)}")
        return super().__new__(cls, raw)

    def hex(self) -> str:
        return bytes(self).hex()


def hash_bytes(data: bytes) -> Hash:
    """Return the 34-byte BLAKE3 multihash of `data`."""
    digest = blake3.blake3(data).digest()
    return Hash(bytes([BLAKE3_CODE, BLAKE3_LEN]) + digest)


def hash_hex(data: bytes) -> str:
    """Return the hex string of the multihash of `data`."""
    return hash_bytes(data).hex()


def from_hex(hexstr: str) -> Hash:
    """Reconstruct a Hash from its hex string."""
    return Hash(bytes.fromhex(hexstr))


def is_valid_hash(h: bytes) -> bool:
    """True iff `h` is a well-formed 34-byte BLAKE3 multihash."""
    return (
        len(h) == HASH_SIZE
        and h[0] == BLAKE3_CODE
        and h[1] == BLAKE3_LEN
    )
